### PR TITLE
Feat/conversation event transcoding capability endpoints

### PR DESCRIPTION
**Pull request by Dovchik**
Covers: conversation/events get, list, delete, send
conversation/conversation inject-event

Add APPLEBC into list of Channels

CI blocked by https://github.com/sinch/doppelganger/pull/13

### CODE REVIEW

File: tests/Sinch.Tests/e2e/Verification/VerificationStatusTests.cs [hunk 1]
```diff
@@ -8,23 +8,24 @@ namespace Sinch.Tests.e2e.Verification
...
-                        Amount = 0.0127
+                        CurrencyId = "some_string_value",
+                        Amount = 1.1
                     }
                 },
-                Reason = Reason.Fraud,
-                Reference = "12345",
-                Source = Source.Intercepted,
-                Status = VerificationStatus.Fail
+                Reason = new Reason("some_string_value"),
```
Comment (asein-sinch) [inline tests/Sinch.Tests/e2e/Verification/VerificationStatusTests.cs:25]: Same comment as in PR#44, whichever comes first

File: src/Sinch/Conversation/Events/Events.cs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    /// </summary>
+    public interface ISinchConversationEvents
+    {
+        /// <summary>
+        ///     Sends an event to the referenced contact from the referenced app. Note that this operation enqueues the event in a queue so a successful response only indicates that the event has been queued.
+        /// </summary>
+        /// <param name="request"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        Task<SendEventResponse> Send(SendEventRequest request, CancellationToken cancellationToken = default);
```
[the remark this answers is not in the document]
Comment (Dovchik) [reply src/Sinch/Conversation/Events/Events.cs:24]: I'll add a task for that

[the remark this answers is not in the document]
Comment (Dovchik) [reply src/Sinch/Conversation/Events/Events.cs:24]: @asein-sinch Can you specify what function signature will add elegancy?

File: src/Sinch/Conversation/Conversations/InjectEvent/InjectEventRequest.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        public string ConversationId { get; set; }
+#endif
+
+        /// <summary>
+        ///     The content of the events.
+        /// </summary>
+#if NET7_0_OR_GREATER
+        public required ConversationEventEvent Event { get; set; }
+#else
+        public ConversationEventEvent Event { get; set; }
```
[the remark this answers is not in the document]
Comment (asein-sinch) [reply src/Sinch/Conversation/Conversations/InjectEvent/InjectEventRequest.cs:26]: yes

File: src/Sinch/Conversation/Events/Events.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (asein-sinch) [reply src/Sinch/Conversation/Events/Events.cs:24]: One method per type of event:

- sendComposingEvent
- sendComposingEndEvent
- sendCommentReplyEvent
- sendAgentJoinedEvent
- sendAgentLeftEvent
- sendGenericEvent
- send

The same can be done for messages btw

File: tests/Sinch.Tests/e2e/Conversation/ConversationsTests.cs [hunk 4]
```diff
@@ -132,6 +135,30 @@ public async Task Stop()
...
+                AcceptTime = DateTime.Parse("1970-01-01T00:00:00Z", CultureInfo.InvariantCulture).ToUniversalTime(),
+                ChannelIdentity = new ChannelIdentity()
+                {
+                    Identity = "01HN31W37910AANG1JGE8Y6RFF",
+                    Channel = ConversationChannel.Instagram,
+                    AppId = "123"
+                },
+                ContactId = _conversation.ContactId,
+                ProcessingMode = ProcessingMode.Dispatch,
+                Event = new ConversationEventEvent(new AppEvent(new ComposingEvent())),
```
[the remark this answers is not in the document]
Comment (Dovchik) [reply tests/Sinch.Tests/e2e/Conversation/ConversationsTests.cs]: Fixed

File: tests/Sinch.Tests/e2e/Verification/VerificationStatusTests.cs (hunk 1 above)
Comment (Dovchik) [reply tests/Sinch.Tests/e2e/Verification/VerificationStatusTests.cs:25]: That's commented on #44

**Comment by Dovchik** [comment]
Latest commit makes InjectMessage to take conversationId as a param and not included in the request object due to system.text.json limitations

File: src/Sinch/Conversation/Conversations/Conversations.cs [hunk 5]
```diff
@@ -244,20 +245,19 @@ public Task Stop(string conversationId, CancellationToken cancellationToken = de
         }
 
         /// <inheritdoc />
-        public Task<InjectEventResponse> InjectEvent(InjectEventRequest injectEventRequest,
+        public Task<InjectEventResponse> InjectEvent(string conversationId, InjectEventRequest injectEventRequest,
```
Comment (asein-sinch) [inline src/Sinch/Conversation/Conversations/Conversations.cs:248]: Why do you need to extract this parameter from `InjectEventRequest` ?

Comment (Dovchik) [reply src/Sinch/Conversation/Conversations/Conversations.cs:248]: The behavior is counterintuitive, If field is marked `required` and `JsonIgnore` at the same type this throws a runtime error from json serializer. I'll include `conversationId` in request class. Also, there will be task to harmonize approach inside all over the solution

Comment (Dovchik) [reply src/Sinch/Conversation/Conversations/Conversations.cs:248]: For the curious - https://github.com/dotnet/runtime/issues/82879

File: src/Sinch/Conversation/Conversations/Conversations.cs [hunk 6]
```diff
@@ -245,19 +244,19 @@ public Task Stop(string conversationId, CancellationToken cancellationToken = de
...
+        public Task<InjectEventResponse> InjectEvent(InjectEventRequest injectEventRequest,
             CancellationToken cancellationToken = default)
         {
             if (injectEventRequest == null)
                 throw new ArgumentNullException(nameof(injectEventRequest), "Shouldn't be null");
 
-            if (string.IsNullOrEmpty(conversationId))
-                throw new ArgumentNullException(nameof(conversationId));
+            if (string.IsNullOrEmpty(injectEventRequest.ConversationId))
+                throw new NullReferenceException($"{nameof(injectEventRequest.ConversationId)} is required.");
```
Comment (asein-sinch) [inline src/Sinch/Conversation/Conversations/Conversations.cs:254]: Nit. the exception message is different from `InjectMessage`

**Review by asein-sinch — APPROVED** [review]

**Pull request merged**